### PR TITLE
Update suggested API.

### PR DESCRIPTION
**Pull request by agl**
This change updates the proposed API to use IdentityCredential instead.

### CODE REVIEW

**Review by davidz25 — COMMENTED** [review]
+1 from me on constructing the request via an API. I also like the fact that it's possible to generate on the server side. Thanks.

File: README.md [hunk 1]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
+  long days;
+};
 
-    required sequence<CredentialElement> requestedElements;
+dictionary IdentityProviderConfig {
+  required string scheme;
 
-    CredentialStorageDuration desiredStorageDuration;  // Not providing this is equivalent to not asking to store.
+  // Common fields across all identity provider types:
+  required DOMString nonce;
```
Comment (othermaciej) [inline README.md:123]: It's hard to follow how this integrates with FedCM or CredentialManager. A comment earlier says this extended a Credential Manager request, but what it links to is only specified in FedCM (afaict) and this `IdentityProviderConfig` definition is incompatible with the one in FedCM. It lacks two of the attributes defined in FedCM, has an extra attribute (`scheme`), and defines one attribute (`nonce`) with a different WebIDL type. What's the actual intent for how this integrates with FedCM and Credential Management? The PR (and ideally the post-PR explainer text) needs to spell it out.

**Review by othermaciej — CHANGES_REQUESTED** [review]

**Review by othermaciej — CHANGES_REQUESTED** [review]
This was difficult to review because the PR does not directly provide motivation, either in PR comments or in the changes to the Explainer document. It essentially deletes and replaces the existing API proposal; it appears to add two other specs as dependencies (Credential Management and FedCM) of which one is itself incubation-stage, making it harder for this proposal to progress and increasing the implementation cost; it makes simple requests for mdoc properties more verbose to express for content authors; and it increases the risk that websites will inappropriately try to use mdoc credentials as a form of user login/authentication, which is highly undesirable. These significant downsides need to be balanced by a strong upside, which is not really given here.

I have since been told there is an older closed PR and an open issue which give some of the motivation, but they aren't even linked here, and the PR and proposed Explainer text should explain these things directly for the benefit of future reviewers of the Explainer.

It was mentioned to me that the motivation is to allow an RP to interchangeably request info from either an mdoc or a FedCM provider as the source of common info (at the user's option). But it's not clear from this PR how these changes would make that use case possible, nor do they add anything to the explained that proposes this as a use case and justifies it as desirable and feasible. As it is, this PR seems like just a pun on two totally different senses of identity - the real-world identity document sense, and the federated login account sense. The ways to request an mdoc or a FedCM credential are divergent, what one gets back needs to be processed differently, and yet the same types are used. These choices need more explanation to evaluate whether this change is a good one.

**Comment by bslassey** [comment]
> I have since been told there is an older closed PR and an open issue which give some of the motivation, 
@othermaciej I believe that PR #3 and issue #2 are the ones that were referenced.

**Comment by agl** [comment]
Thank you for the feedback!

This change was intended to be a quick aid to some ongoing discussions, but it has been hanging around longer than anticipated, so I apologize that the context was missing. I’ll try to fill it in:
&nbsp;

We also want to see a clear separation between authentication and identity. While we cannot stop identity mechanisms from being used for authentication, we want that to be cutting against the grain. We started by putting this functionality into a new `navigator.identity` namespace, but ultimately decided that `navigator.credentials` was a better fit because of federation.

We would like federation to evolve into an identity mechanism, and for authentication needs to be served by passkeys. So we envision that a site would create an account by creating a passkey and then optionally attach identity attributes to that account, like an email address, by using federation. That's why FedCM currently uses <tt>navigator.credentials.get({_identity_: …})</tt>. But then if we put real-world identity functions in a `navigator.identity` namespace, we would have `identity` appearing in two different places. We felt that undermined the benefits of separation.

Federation is used for both authentication and identity today. We can't separate that in the short-term. And we do want to be able to support authentication requests which accept either a password, or passkey, or federated sign-in. (There are numerous examples of sites that have such a UI right now, although perhaps without the passkey option yet.)

So that's why we put the API into `navigator.credentials`.
&nbsp;

As for tying it with FedCM: as mentioned, we would like to move federation towards being an identity mechanism and so we want to support sites being able to request identity attributes using either mDLs or federation. If mDLs are only ever used for mobile driver's licenses, that doesn't make a lot of sense. But we expect that the mDL data model will be inevitably used for other things too. For example, proving an academic affiliation is often implemented via a federation-like mechanism today (SAML, specifically), but it seems reasonable that it could be an mdoc too (e.g. University Certificates). (And the mdoc would have better privacy properties.) So an online research library might reasonably request proof of affiliation, but want to accept either mechanism.
&nbsp;

While it is nice to be able to make invalid requests, like requesting a disjunction of an authentication and identity mechanism, inexpressible, it is only one method of communicating with developers. We expect that browsers would immediately reject any requests that ask for either a passkey or an mdoc (for example), and that communicates the separation pretty clearly too. There is also documentation, language in specs, established norms, etc to communicate the separation.
&nbsp;

Another facet of the design comes from our experience with Web Authentication / passkeys. Web developers really want to be able to pass authentication requests/responses easily to the server, but WebAuthn makes frequent use of ArrayBuffers, which don’t work with JSON.stringify and so do not easily serialize into JSON. Thus one of the goals of the design here is that the request be in the subset of JavaScript that works with standard JSON functions.
&nbsp;

We do propose that this spec be built on top of the IdentityCredential requested through the “identity” option of the Credential Manager API (of which FedCM is built on top of), but not in a way that requires an implementation to also implement FedCM.

This PR was only meant to be a sketch. As you note, it duplicates some bits from other specs where it was clearer to do so, rather than to refer to those documents. Ultimately, all these cases have to be fixed, but at the time we didn't want to distract from the core points. Putting retention as a top-level member was a mistake, one that we already realized some weeks ago, and were intending to address in a follow-up. It seemed like it might be a concept that spans multiple identity mechanisms, but we now feel that keeping mechanisms completely separate is probably best right now.

Speaking of multiple mechanisms, another goal that we have is to ensure that there is space in the API for requesting Verifiable Credentials. We are focused on building an mDL API, but the possibility that Verifiable Credentials will be an ecosystem necessity in the future is real, and we want to ensure that should they need to be supported, it would not damage the API too much.
&nbsp;

So that's why it feels to us like we'll be reducing the risk of having a mess on our hands long term if we put this into `navigator.credentials`. But as mDL is rapidly gaining traction, we feel some urgency to pick a design so that we can work towards prototyping a first version in the coming months.

Does it seem reasonable if we open an issue in this repo, dedicated to this question, and ask other ecosystem members (including TAG) to weigh in?

**Comment by othermaciej** [comment]
@agl can you put more of what’s in your comment into the explainer draft in your PR, as well as addressing all the specific points of feedback of bits that are unclear or seem like outright errors? The intent of making a sketch is reasonable, but my understanding is that the Apple folks who originally proposed this spec have been asked to review and merge the PR, and I do not think it is mergeable in its current form.

implied requirements like rejecting any credential request for either a passkey or an mDL should be spelled out, I think. As should details like proposals to change things in FedCM, or to push bits of FedCM down to CM. It was impossible for me to tell which cases were things like that vs simply errors/oversights.

As for an interchangeable request for either an mdoc or a FedCM credential, I still do not understand even with your explanation how that is supposed to work. If it’s supposed to work with this PR and FedCM as it stands, then adding a worked example to the explainer would help. If it’s not supposed to work as-is and would require changes to FedCM proper, then sketching those changes would help. Moving this API to Credential Management has some clear downsides (more verbosity, makes it possible to express invalid requests) so the proposed API shape needs to be understandable even to those who do not have the inside thinking of how FedCM is expected to evolve. So I think more of the context above needs to be in the Explainer and the path to joint use with FedCM needs to be spelled out with more specifics, ideally a code examaple.

I think that, plus fixing the technical errors highlighted in my specific line comments, is a reasonable request before a new round of review of this PR.

**Comment by RByers** [comment]
@othermaciej, it seems like the key design question to be answered (before spending time on a lot of these fine details) is whether unifying with CM is a good idea or not, right? We've taken your feedback here and updated the discussion in #2, shall we continue there?

I've also updated this PR to align and will respond to each comment individually, but perhaps we should see if we can get some consensus on #2 first before worrying too much about the details here?

File: README.md [hunk 2]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
 
 ### API
 
-One idea a new `CredentialRequest` (et al)
-```
-dictionary CredentialElement {
-    required DOMString namespace;  // As defined in ISO 18013-5 clause 8.
-    required DOMString name;
+We propose to build upon the Credential Manager’s “identity” request, which [Federated Credential Management](https://fedidcg.github.io/FedCM/#dictdef-identitycredentialrequestoptions) is also built on top of, with the following structures:
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: We've started an explainer for the common pieces [here](https://github.com/agl/identity-credential/blob/main/README.md). We could incubate this separately in WICG if there's interest?

File: README.md [hunk 3]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
+
+```webidl
+// The existing extension to the credential management API:
+partial dictionary CredentialRequestOptions {
+  IdentityCredentialRequestOptions identity;
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: Fixed by [IdentityCredential](https://github.com/agl/identity-credential/blob/main/README.md)

File: README.md [hunk 4]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
-    // At least one of these is required.
+dictionary IdentityCredentialRequestOptions {
+  required sequence<IdentityProviderConfig> providers;
 
-    boolean forever;  // Cannot be used with any other properties.
+  // Common fields across all types of providers:
 
-    long days;  // Cannot (currently) be used with any other properties.
+  // Omitting `retention` implies that the identity information will not be stored.
+  IdentityStorageDuration retention;
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: Fixed in latest PR. At the moment we're proposing keeping retention as an mdoc-specific thing from a desire to not go too far in trying to force things together as they're still evolving.

[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: Agreed, let's keep retention mdoc-specific for now.

File: README.md [hunk 5]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
 };
 
-dictionary CredentialDocumentDescriptor {
-    required DOMString documentType;  // As defined in ISO 18013-5 clause 8.
+dictionary IdentityStorageDuration {
+  // Exactly one of the following must be provided.
+  boolean forever;
+  long days;
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: Agreed thats simpler, done.

File: README.md [hunk 6]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
-    CredentialStorageDuration desiredStorageDuration;  // Not providing this is equivalent to not asking to store.
+  // Common fields across all identity provider types:
+  required DOMString nonce;
 };
 
-dictionary CredentialDocument {
-    object data;  // The CBOR encoded `CredentialDocument` defined above.
+// extends the IdentityProviderConfig with mDoc-specific attributes
+partial dictionary IdentityProviderConfig  {
+  // For use where `scheme` is `mdoc`:
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: To be honest, I think it's hard to be sure how the major use cases like age verification will play out at this stage. Will federated IDPs want to offer such a service along side the option of using government-issued IDs? I, for one, can totally imagine users being more comfortable with letting, say, Apple attest to their age rather than share the fact that their California driver's license says they're over 21. 

[Here's an example](https://github.com/agl/identity-credential/blob/main/README.md#reconcilation) that we think is a bit less uncertain at this stage - proving your University affiliation using either FedCM or an mdoc.

File: README.md [hunk 7]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
+  // base64-encoded reader certification chain, if the platform requires it.
+  optional sequence<DOMString> certificates;
+
+  required DOMString documentType;
+  required sequence<MdocElement> requestedElements;
+};
+
+dictionary MdocElement {
+  required DOMString namespace;  // As defined in ISO 18013-5 clause 8.
+  required DOMString name;
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: I don't think it's actually important that mdoc and FedCM share specific properties or any aspect of their query language (though maybe they will in the future). I think what's more important is that we don't create a design where the UI is necessarily coupled 1:1 to the specific technology choice. Websites shouldn't have to have "how do you want to prove your identity: mdoc wallet, VC wallet, google account, ..." buttons.

File: README.md [hunk 8]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
 };
 
-dictionary RequestConfiguration {
-    required DOMString nonce;
+partial dictionary IdentityProviderConfig {
+  // ...
+  // Federated provider specific configs, not covered here.
+  // ...
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: IMHO the main benefit is being able to create an ordered list of different possible sources of identity information so that the browser can construct a meaningful UI that (potentially) merges information from more than one source.

File: README.md [hunk 9]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
 };
+```
 
-[
-    SecureContext,
-    Exposed=Window,
-] interface CredentialRequest {
-    constructor(DOMString requesterIdentity, CredentialDocumentDescriptor documentDescriptor);  // This throws if anything in the `documentDescriptor` is not recognized (e.g. an invalid `documentType`).
+There is always a bijection from `IdentityCredentialRequestOptions` to JSON strings, and that bijection is implemented by `JSON.parse` and `JSON.stringify`. This allows requests to be generated on the server and easily transported to the user agent.
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: Agreed, fixed

File: README.md [hunk 10]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
 
-    Promise<CredentialDocument> requestDocument(RequestConfiguration configuration);
+When an `IdentityCredentialRequestOptions` is used with [`get`](https://www.w3.org/TR/credential-management-1/#dom-credentialscontainer-get), the result is an `IdentityCredential`:
 
-    Promise<undefined> abort();
+```webidl
+[Exposed=Window, SecureContext]
+interface IdentityCredential : Credential {
+  // The CBOR encoded `CredentialDocument` defined above, wrapped in base64.
+  DOMString token;
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: We imagine that identity information will be passed between various pieces of tech stack where only the endpoints need to be able to parse and understand it. For example, in Chrome we currently plan to pass any information provided by the platform wallet through without parsing it ourselves. Website front ends may do the same, just passing the response on to some server component. 

Agreed though that `token` is inappropriate. In our latest `IdentityCredential` [proposal](https://github.com/agl/identity-credential/blob/main/README.md), we use an opaque `response`.

File: README.md [hunk 11]
```diff
@@ -93,80 +93,125 @@ EncryptionParamaters = {
...
 };
 ```
 
+The `id` of the `IdentityCredential` for an mdoc response would be `mdoc`.
```
[the remark this answers is not in the document]
Comment (RByers) [reply README.md]: Agreed, fixed

File: README.md [hunk 12]
```diff
@@ -15,47 +16,115 @@
...
+
+```js
+let request = {
+  identity: {
+    providers: [{
+      mdoc: {
+        nonce: "gf69kepV+m5tGxMyMF0cnn9NCnRHez/LUIsFtLi6pwg=",
+        documentType: "org.iso.18013.5.1.mDL",
+        retentionDays: 90,
+        readerPublicKey: ["ftl+VEHPB17r2oi6it3ENaqhOOB0AZbAkb5f4VlCPakpdNioc9QZ7X/6w..."],
```
Comment (snorp) [inline README.md:102]: This looks base64-encoded, but do we know any more than that? Is it a SEC 1 uncompressed EC key? Something else?
